File: listen.rb
#!/usr/bin/env ruby -wW1

# listen.rb is a simple script that receives rabbit events as JSON
# blobs, decodes and pretty prints them on teh console.

require 'amqp'
require 'json'

EventMachine.run do
  connection = AMQP.connect(:host => '127.0.0.1')
  puts "Connected to AMQP broker. Running #{AMQP::VERSION} version of the gem..."

  channel  = AMQP::Channel.new(connection)
  exchange = channel.fanout("rabbitevents")
  channel.queue("rabbitevents-#{Process.pid}", :auto_delete => true).bind(exchange).subscribe do |headers, payload|
    puts "[event] Received a message"
    if headers.content_type && headers.content_type == 'application/json'
      puts JSON.pretty_generate(JSON.parse(payload))
    else
      puts "        headers: #{headers.attributes}"
      puts "        payload: #{payload.inspect}"
    end
  end

end

File: listen.erl
#!/usr/bin/env escript 
%%! -pz ./amqp_client ./rabbit_common ./amqp_client/ebin ./rabbit_common/ebin

-include_lib("amqp_client/include/amqp_client.hrl").

main(_) ->
    {ok, Connection} =
        amqp_connection:start(#amqp_params_network{host = "localhost"}),
    {ok, Channel} = amqp_connection:open_channel(Connection),

    amqp_channel:call(Channel, #'exchange.declare'{exchange = <<"rabbitevents">>,
                                                   type     = <<"fanout">>}),

    #'queue.declare_ok'{queue = Queue} =
        amqp_channel:call(Channel, #'queue.declare'{exclusive = true}),

    amqp_channel:call(Channel, #'queue.bind'{exchange = <<"rabbitevents">>,
                                             queue    = Queue}),

    io:format(" [*] Waiting for events. To exit press CTRL+C~n"),

    amqp_channel:subscribe(Channel, #'basic.consume'{queue = Queue,
                                                     no_ack = true}, self()),
    receive
        #'basic.consume_ok'{} -> ok
    end,
    loop(Channel).

loop(Channel) ->
    receive
        {#'basic.deliver'{}, #amqp_msg{payload = Body}} ->
            io:format(" got an event", []),
            Json = mochijson2:decode(Body),
            io:format(" [x] ~p~n", [Json]),
            loop(Channel)
    end.
